Add pages/recipes for common questions (#3106)
cerbos/cerbos: docs/modules/recipes/pages/filtering-resources.adoc

  "principal": {
    "id": "alice",
    "roles": ["user"],
    "attr": {
      "department": "engineering"
    }
  },
  "resource": {
    "kind": "document",
    "policyVersion": "default"
  }
}
----

== Plan result kinds

The response contains one of three result kinds:

`ALWAYS_ALLOWED`:: The principal can access all resources of this kind for this action. No filtering is needed.

`ALWAYS_DENIED`:: The principal cannot access any resources of this kind. Return an empty result set.

`CONDITIONAL`:: The response includes a filter condition as an AST. Convert this to a database query to fetch only the resources the principal can access.

.Example CONDITIONAL response
[source,json,linenums]
----
{
  "requestId": "filter-example",
  "action": "view",
  "resourceKind": "document",
  "filter": {
    "kind": "KIND_CONDITIONAL",
    "condition": {
      "expression": {
        "operator": "eq",
        "operands": [
          { "variable": "request.resource.attr.department" },
          { "value": "engineering" }
        ]
      }
    }
  }
}
----

In this example, the plan tells your application to filter documents where `department = "engineering"`.

== Query plan adapters

> Converting the AST into database queries by hand can be tedious. Cerbos provides reference adapter implementations for common data layers that handle this conversion for you.

Reviewer: ```suggestion
Cerbos provides reference adapter implementations for common data layers as guidance for your own implementations.
```
